### PR TITLE
Implement ConEmu's OSC 9;9 to set the CWD

### LINKED ISSUE #8166: Consider using OSC 9;9 to inform Windows Terminal about CWD
# Description of the new feature/enhancement

About a month ago I've filed #7668 , in the seek of a solution for #3158 . It raised a lot of discussion and I can totally understand why #7668 is acceptable to some people but unpalatable to others. 

Recently with #8055 being accepted as a solution, I thought why not accept more sequences from ConEmu, which is a fantastic product I've personally used for a long time. So here I purpose we implement another OSC 9 custom sequence from ConEmu:

|Sequence|Descriptoin|
| :------------- | :----------: |
|ESC ] 9 ; 9 ; “cwd” ST | Inform ConEmu about shell current working directory.|

The source can be found in the [ConEmu doc](https://conemu.github.io/en/AnsiEscapeCodes.html#ConEmu_specific_OSC).

# Proposed technical implementation details (optional)

PR: #8330

Shell configurations:
* PowerShell: https://gist.github.com/skyline75489/480d036db8ae9069b7009377e6eebb79
* WSL: https://gist.github.com/skyline75489/d655aede4c729eff178a1c0bfd10f622

### PR DESCRIPTION
**Pull request by skyline75489**
## Summary of the Pull Request

This PR implement the OSC 9;9 

|Sequence|Descriptoin|
| :------------- | :----------: |
|ESC ] 9 ; 9 ; “cwd” ST | Inform ConEmu about shell current working directory.|

 
## References

#8214

## PR Checklist
* [X] Closes #8166
* [X] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

### CODE REVIEW

**Comment by skyline75489** [comment]
So this is basically a reborn version of #7668 . Thanks to prior work of @PankajBhojwani, this PR is even smaller than #7688 .

Based on previous thread like #3158 and #8214, I'd like to implement OSC 9;9 as a "Windows-first" sequence that will assume users will set Windows path . This can be seen as a starter towards a more comprehensive implementation of OSC 7, which will handle all kinds of paths based on the profile used.
 
As for this PR, it gives users the ability to make use the of the CWD when duplicating the current active tab. I don't really know how to handle splitting panes so I left it unchanged.

I'd like to hear you guys opinions.

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 1]
```diff
@@ -1015,6 +1015,13 @@ namespace winrt::TerminalApp::implementation
                     if (profileGuid.has_value())
                     {
                         const auto settings{ winrt::make<TerminalSettings>(_settings, profileGuid.value(), *_bindings) };
+                        const auto workingDirectory = terminalTab->GetActiveTerminalControl().WorkingDirectory();
+                        const auto validWorkingDirectory = !workingDirectory.empty();
```
Comment (skyline75489) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1019]: I relaxed the validation per the comments by @zadjii-msft in https://github.com/microsoft/terminal/issues/8166#issuecomment-724678264

I'm not sure how we handle `wt -d c:\some\invalid\path` nowadays. Could use some help here.

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1019]: So currently, `wt -d c:\some\invalid\path` will do this:

image


That's not a super useful error, `ERROR_DIRECTORY` - I guess that makes sense, though not totally obvious what's wrong here. 

I'm okay with us leaving this for now. We do some validation of the `startingDirectory` on settings load, but nothing for directories passed as commandline args, and I think that's fine.

If we did want to try and pre-validate, we could pull `Profile::EvaluateStartingDirectory` into a helper, and only use the control's directory if it's valid. That's actually probably a good idea.

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
Overall, this looks pretty good. No major concerns - but I'm on the fence about validating the paths ahead of time. _I want this to land_, but maybe we should validate ahead of time.

I'll block for the discussion, but it's otherwise a fine PR

EDIT: I'm also blocking over the pane duplicating. We should probably do that.

**Comment by zadjii-msft** [comment]
As far as pane duplicating goes, that's down in `TerminalPage::_SplitPane`. Somewhere near:

```c++
            if (splitMode == SplitType::Duplicate)
            {
                std::optional<GUID> current_guid = focusedTab->GetFocusedProfile();
                if (current_guid)
                {
                    profileFound = true;
                    controlSettings = { winrt::make<TerminalSettings>(_settings, current_guid.value(), *_bindings) };
                    realGuid = current_guid.value();
                }
```

**Comment by skyline75489** [comment]
Another thing that I'm not so sure: in #7668 I make both `new-tab` and `duplicate-tab` use the CWD. In this PR, I figured it's too aggresive so I changed it to just `duplicate-tab`. Is this too aggresive, or what people would want? Personally I'm OK with `new-tab` also accepting CWD but I'm not sure what others will feel about this.

**Comment by zadjii-msft** [comment]
I'd say let's just start with the `duplicateTab` action. We could probably consider adding an argument to `newTab` to allow it to duplicate paths as well. That's my opinion at least

**Comment by skyline75489** [comment]
I've been thinking about the path validation, then it occurs to me that the OSC 9;9 feature is currently a super opt-in kind of things. Users will need to modify their shell script in order to even get this up and running. So I guess even if they have wrongly configured the path and be frustrated by the `error when launching x.exe`, they are likely to know the root cause of it is exactly the misconfiguration created by themselves. Besides, with only `duplicateTab/Pane` enabled now, there's little chance that a WSL path is passed to `powershell.exe` or vice versa, which considerably reduces the chances of  misformated URLs being passed around across different profiles. So yeah I think this is readly now.

**Comment by zadjii-msft** [comment]
Yea I'm fine with not path validating as well. Can we get this added to duplicating panes though too 😄?

**Review by zadjii-msft — APPROVED** [review]
I'm a derp, this does work with pane splitting too. So let's do this

**Review by carlos-zamora — APPROVED** [review]

**Comment by tavrez** [comment]
I can't use this in todays releases anymore(both stable and preview). Somethig changed?

**Comment by skyline75489** [comment]
@tavrez No nothing changed I believe. Are you positive that you are using the "duplicate tab"(ctrl + shift + d) instead of "creating new tab" (ctrl + shift + t) ? If you have a consistent repro, feel free to open a new issue.

**Comment by tavrez** [comment]
Yes I'm using duplicate tab(with keyboard shortcut and from command pallete), It doesn't work, it goes back to starting directory. If you are not having this problem(simply open a new terminal, change directory, press duplicate tab) probably I'm doing something wrong(but this procedure worked for me in prev version it introduced in Terminal Preview)

**Comment by skyline75489** [comment]
@tavrez It's working for me. Sorry I guess you'll need to check the shell configuration you're using.

**Comment by LeaveTheCapital** [comment]
This does not work for me either in latest preview, or in latest Windows Terminal, when using duplicate tab (ctrl + shift + d). Am I missing something? Is there a setting that needs to be enabled?

**Comment by tavrez** [comment]
@LeaveTheCapital just to make sure, do you send the sequence code (ESC ] 9 ; 9 ; “cwd” ST) before duplicating? If yes, you probably has a weird problem like me, I can't use this only on my personal PC

**Comment by LeaveTheCapital** [comment]
> @LeaveTheCapital just to make sure, do you send the sequence code (ESC ] 9 ; 9 ; “cwd” ST) before duplicating? If yes, you probably has a weird problem like me, I can't use this only on my personal PC

I did try this, however I must have been doing something wrong - perhaps someone could help with an idiots guide, or a link to similar? Thank you

**Comment by skyline75489** [comment]
@LeaveTheCapital You can try the [gist](https://gist.github.com/skyline75489/480d036db8ae9069b7009377e6eebb79) to see if it helps. Make sure you are using "Duplicate Tab" (Ctrl+Shift+d, by default) and not "New Tab" (Ctrl + T)

**Pull request merged**